src/trap_routines.rs

```diff
@@ -0,0 +1,272 @@
+    Ok(())
+}
+
+/// Output a single character.
+pub fn out(regs: &mut Registers, writer: &mut impl Write) -> Result<(), VMError> {
+    let c: u8 = regs[Register::R0]
+        .try_into()
+        .map_err(|e: TryFromIntError| VMError::Conversion(e.to_string()))?;
+    stdout_write(&[c], writer)?;
+    Ok(())
+}
+
+/// Prompt for input character from the user.
+pub fn trap_in(
+    regs: &mut Registers,
+    writer: &mut impl Write,
+    reader: &mut impl Read,
+) -> Result<(), VMError> {
+    print!("Enter a character: ");
+    let buffer = getchar(reader)?;
+    stdout_write(&buffer, writer)?;
+    stdout_flush(writer)?;
+    regs[Register::R0] = buffer[0].into();
+    update_flags(Register::R0, regs);
+    Ok(())
+}
+
+/// Output a null-terminated string. The characters are contained in consecutive memory locations,
+/// one character per memory location, starting with the address specified in R0. Writing
+/// terminates with the occurrence of x0000 in a memory location.
+pub fn puts(
+    regs: &mut Registers,
+    mem: &mut Memory,
+    writer: &mut impl Write,
+) -> Result<(), VMError> {
+    // Get the address of the first character and read it
+    let mut c_addr = regs[Register::R0];
+    let mut c = mem.read(c_addr)?;
+    while c != NULL {
+        // Parse it into a u8, write it and pass to the next memory location
+        let char: u8 = c
+            .try_into()
+            .map_err(|e: TryFromIntError| VMError::Conversion(e.to_string()))?;
+        stdout_write(&[char], writer)?;
+        c_addr = c_addr.wrapping_add(1);
+        c = mem.read(c_addr)?;
+    }
+    stdout_flush(writer)?;
+    Ok(())
+}
+
+/// Output a null-terminated string. The characters are contained in consecutive memory locations,
+/// but this time there are two characters per memory location, starting with the address specified in R0. Writing
+/// terminates with the occurrence of x0000 in a memory location.
+pub fn puts_p(
+    regs: &mut Registers,
+    mem: &mut Memory,
+    writer: &mut impl Write,
+) -> Result<(), VMError> {
+    // Get the address of the first characters and read them
+    let mut c_addr = regs[Register::R0];
+    let mut c = mem.read(c_addr)?;
+    while c != NULL {
+        // Get the first character in the memory location (the 8 leftmost bits)
+        let char1 = (c & 0xFF)
+            .try_into()
+            .map_err(|e: TryFromIntError| VMError::Conversion(e.to_string()))?;
+        stdout_write(&[char1], writer)?;
+        // Get the second character in the same memory location (the 8 rightmost bits)
+        let char2 = (c >> 8)
+            .try_into()
+            .map_err(|e: TryFromIntError| VMError::Conversion(e.to_string()))?;
+        if char2 != 0x00 {
+            stdout_write(&[char2], writer)?;
+        }
+        c_addr = c_addr.wrapping_add(1);
+        // Get the next memory location
+        c = mem.read(c_addr)?;
+    }
+    stdout_flush(writer)?;
+    Ok(())
+}
+
+/// Halt program
```

Comment:
What are the implications of halting the program? I think this can be documented in two ways:

- Document the implications of halting the program in this function. This means the documentation should point to the function that calls this to add more context.
- Document the implications of halting the program in the caller function.